aten/src/ATen/native/sparse/SparseUnaryOps.cpp

```diff
@@ -0,0 +1,147 @@
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+
+namespace at {
+namespace native {
+namespace {
+
+template <typename Ufunc>
+Tensor coalesced_unary_ufunc(const Tensor &self, const Ufunc &ufunc) {
+  TORCH_INTERNAL_ASSERT(self.is_sparse());
+  const auto input = self.coalesce();
+  Tensor out_values = ufunc(input.values());
+  Tensor result = at::_sparse_coo_tensor_with_dims_and_tensors(
+      input.sparse_dim(),
+      input.dense_dim(),
+      input.sizes(),
+      input.indices().clone(),
+      out_values,
+      input.options().dtype(out_values.scalar_type()));
+  result._coalesced_(true);
+  return result;
+}
+
+template <typename Ufunc>
+Tensor& coalesced_unary_ufunc_(Tensor &self, const Ufunc &ufunc) {
+  TORCH_INTERNAL_ASSERT(self.is_sparse());
+  auto values = self._values();
+  ufunc(values);
+  return self;
+}
+
+template <typename Ufunc>
+Tensor& coalesced_unary_ufunc_out(const Tensor &self, Tensor &result, const Ufunc &ufunc) {
+  if (self.is_same(result)) {
+    TORCH_CHECK(self.is_coalesced(), "expected coalesced tensor for inplace operation");
+    auto values = self._values();
+    ufunc(values, values);
+    return result;
+  }
+
+  TORCH_CHECK(self.is_sparse() && result.is_sparse());
+  const auto input = self.coalesce();
+  sparse_resize_(result, input.sizes(), input.sparse_dim(), input.dense_dim());
+
+  auto input_values = input.values();
+  auto result_values = result._values();
+  result_values.resize_(input_values.sizes());
```

Comment:
Resizing indices&values of sparse tensors doesn't work like this. In this case, `result_values` is resized, but `result._values()` still has the original size. You need to use `SparseTensorImpl::set_indices_and_values_unsafe` to really set the newly resized tensors as indices and values. See also https://github.com/pytorch/pytorch/issues/63549

```python
In [1]: import torch

In [2]: a = torch.zeros(3, 3).to_sparse()

In [3]: a_indices = a._indices()

In [4]: a.shape
Out[4]: torch.Size([3, 3])

In [5]: a_indices.shape
Out[5]: torch.Size([2, 0])

In [6]: a_indices.resize_((2, 3))
Out[6]:
tensor([[140315210074560, 140315191773632, 140315210132992],
        [140315210045888, 140315210131968, 140315210430464]])

In [7]: a_indices.shape
Out[7]: torch.Size([2, 3])

In [8]: a._indices().shape
Out[8]: torch.Size([2, 0])
```

Reply:
I've changed it to access `values` and `indices` without dispatching, will that work?

Reply:
I guess it should work now. Best to have a test passing "out=" tensor with the wrong size to check.